flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java

```diff
@@ -0,0 +1,188 @@
+     * thread safety should be ensured.
+     */
+    private final Deque<Tuple2<Buffer, Integer>> allBuffers = new LinkedList<>();
+
+    private CompletableFuture<Void> flushCompletableFuture = FutureUtils.completedVoidFuture();
+
+    /**
+     * Record the segment id that is writing to.
+     *
+     * <p>Note that when flushing buffers, this can be touched by task thread or the flushing
+     * thread, so the thread safety should be ensured.
+     */
+    @GuardedBy("allBuffers")
+    private int segmentId = -1;
+
+    /**
+     * Record the buffer index in the {@link SubpartitionRemoteCacheManager}. Each time a new buffer
+     * is added to the {@code allBuffers}, this field is increased by one.
+     *
+     * <p>Note that the field can only be touched by the task thread, so this field need not be
+     * guarded by any lock or synchronizations.
+     */
+    private int bufferIndex;
+
+    public SubpartitionRemoteCacheManager(
+            TieredStoragePartitionId partitionId,
+            int subpartitionId,
+            TieredStorageMemoryManager storageMemoryManager,
+            PartitionFileWriter partitionFileWriter) {
+        this.partitionId = partitionId;
+        this.subpartitionId = subpartitionId;
+        this.partitionFileWriter = partitionFileWriter;
+        storageMemoryManager.listenBufferReclaimRequest(this::flushBuffers);
+    }
+
+    // ------------------------------------------------------------------------
+    //  Called by RemoteCacheManager
+    // ------------------------------------------------------------------------
+
+    void startSegment(int segmentId) {
+        synchronized (allBuffers) {
+            this.segmentId = segmentId;
+        }
+    }
+
+    void addBuffer(Buffer buffer) {
+        Tuple2<Buffer, Integer> toAddBuffer = new Tuple2<>(buffer, bufferIndex++);
+        synchronized (allBuffers) {
+            allBuffers.add(toAddBuffer);
+        }
+    }
+
+    void finishSegment(int segmentId) {
+        synchronized (allBuffers) {
+            checkState(this.segmentId == segmentId, "Wrong segment id.");
+        }
+        int numFlushedBuffers = flushBuffers();
+        if (numFlushedBuffers > 0) {
+            PartitionFileWriter.SubpartitionBufferContext bufferContext =
+                    new PartitionFileWriter.SubpartitionBufferContext(
+                            subpartitionId,
+                            Collections.singletonList(
+                                    new PartitionFileWriter.SegmentBufferContext(
+                                            segmentId, Collections.emptyList(), true)));
+            // Notify the partition file writer that the segment is finished through writing the
+            // buffer context
+            flushCompletableFuture =
+                    partitionFileWriter.write(
+                            partitionId, Collections.singletonList(bufferContext));
+        }
```

Comment:
This doesn't seem right. `numFlushedBuffers == 0` doesn't mean there's no data in the segment. It is possible that the buffers are flushed before calling `finishSegment`. E.g., on buffer reclaim request.

Reply:
Fixed.